python: kvs still assumes all KVS values must be json

Per comments in #5037 and #5196.  json assumption was largely removed like in 2016 or something.  Since most users use the  "job" python API, this issue has lingered in the Python KVS functions for ages.

Example from #5196:

```
import flux
import flux.kvs

f = flux.Flux()
d = flux.kvs.KVSDir(f)

print(d['a']['b']['c'])
```

```
>flux kvs put a.b.c=foo
>flux python ./test.py
Traceback (most recent call last):
  File "./test.py", line 7, in <module>
    print(d['a']['b']['c'])
  File "/g/g0/achu/chaos/git/flux-framework/flux-core/src/bindings/python/flux/kvs.py", line 175, in __getitem__
    return get(self.fhdl, self.key_at(key))
  File "/g/g0/achu/chaos/git/flux-framework/flux-core/src/bindings/python/flux/kvs.py", line 71, in get
    return get_key_direct(flux_handle, key)
  File "/g/g0/achu/chaos/git/flux-framework/flux-core/src/bindings/python/flux/kvs.py", line 38, in get_key_direct
    ret = json.loads(ffi.string(valp[0]).decode("utf-8"))
  File "/usr/lib64/python3.6/json/__init__.py", line 354, in loads
    return _default_decoder.decode(s)
  File "/usr/lib64/python3.6/json/decoder.py", line 339, in decode
    obj, end = self.raw_decode(s, idx=_w(s, 0).end())
  File "/usr/lib64/python3.6/json/decoder.py", line 357, in raw_decode
    raise JSONDecodeError("Expecting value", s, err.value) from None
json.decoder.JSONDecodeError: Expecting value: line 1 column 1 (char 0)

>flux kvs put a.b.c=\"foo\"
>flux python test.py
foo
```

I tried the dumbest thing I could think of:

```
diff --git a/src/bindings/python/flux/kvs.py b/src/bindings/python/flux/kvs.py
index 377fd1c..dc6b30d 100644
--- a/src/bindings/python/flux/kvs.py
+++ b/src/bindings/python/flux/kvs.py
@@ -35,7 +35,7 @@ def get_key_direct(flux_handle, key):
     if valp[0] == ffi.NULL:
         return None
 
-    ret = json.loads(ffi.string(valp[0]).decode("utf-8"))
+    ret = ffi.string(valp[0]).decode("utf-8")
     RAW.flux_future_destroy(future)
     return ret
```

The failures in the tests were like:

```
FAIL: python/t0005-kvs.py 6 __main__.TestKVS.test_fill
  ---
    message: |
      Traceback (most recent call last):
        File "./python/t0005-kvs.py", line 94, in test_fill
          self.assertEqual(kd["things"], 1)
      AssertionError: '1' != 1
  ...
```

i.e. there are typing issues that `json.loads()` previously handled.  So then I tried the second stupidest thing I could think of:

```Python
diff --git a/src/bindings/python/flux/kvs.py b/src/bindings/python/flux/kvs.py
index 377fd1c..6e47400 100644
--- a/src/bindings/python/flux/kvs.py
+++ b/src/bindings/python/flux/kvs.py
@@ -35,7 +35,10 @@ def get_key_direct(flux_handle, key):
     if valp[0] == ffi.NULL:
         return None
 
-    ret = json.loads(ffi.string(valp[0]).decode("utf-8"))
+    try:
+        ret = json.loads(ffi.string(valp[0]).decode("utf-8"))
+    except json.decoder.JSONDecodeError:
+        ret = ffi.string(valp[0]).decode("utf-8")
     RAW.flux_future_destroy(future)
     return ret
```

And all tests pass (should add more of course) ... and I might need to handle a few other corner cases (like empty "no value" comes to mind).

To me this seems reasonable.  If it can't be decoded by json.loads(), just assume its a string??  Obviously there's some corner cases here, like if the user were to write binary data.  But this seems an acceptable compromise?  Should also handle the case I hit in #5037.

--- a/src/bindings/python/flux/kvs.py
+++ b/src/bindings/python/flux/kvs.py
@@ -35,7 +35,12 @@ def get_key_direct(flux_handle, key):
     if valp[0] == ffi.NULL:
         return None
 
-    ret = json.loads(ffi.string(valp[0]).decode("utf-8"))
+    try:
+        ret = json.loads(ffi.string(valp[0]).decode("utf-8"))
+    except json.decoder.JSONDecodeError:
+        ret = ffi.string(valp[0]).decode("utf-8")
+    except UnicodeDecodeError:
+        ret = ffi.string(valp[0])
     RAW.flux_future_destroy(future)
     return ret
 
@@ -78,10 +83,17 @@ def get(flux_handle, key):
 
 
 def put(flux_handle, key, value):
-    json_str = json.dumps(value)
     if flux_handle.aux_txn is None:
         flux_handle.aux_txn = RAW.flux_kvs_txn_create()
-    return RAW.flux_kvs_txn_put(flux_handle.aux_txn, 0, key, json_str)
+    try:
+        json_str = json.dumps(value)
+        return RAW.flux_kvs_txn_put(flux_handle.aux_txn, 0, key, json_str)
+    except TypeError:
+        if isinstance(value, bytes):
+            return RAW.flux_kvs_txn_put_raw(
+                flux_handle.aux_txn, 0, key, value, len(value)
+            )
+        raise TypeError
 
 
 def put_mkdir(flux_handle, key):

--- a/t/python/t0005-kvs.py
+++ b/t/python/t0005-kvs.py
@@ -35,10 +35,13 @@ class TestKVS(unittest.TestCase):
         kd = flux.kvs.KVSDir(self.f)
         kd[key] = value
         kd.commit()
-        nv = kd[key]
+
+        kd2 = flux.kvs.KVSDir(self.f)
+        nv = kd2[key]
         self.assertEqual(value, nv)
         self.assertTrue(isinstance(nv, type))
-        return kd
+
+        return kd2
 
     def test_set_int(self):
         self.set_and_check_context("int", 10, int)

--- a/t/python/t0005-kvs.py
+++ b/t/python/t0005-kvs.py
@@ -38,8 +38,12 @@ class TestKVS(unittest.TestCase):
 
         kd2 = flux.kvs.KVSDir(self.f)
         nv = kd2[key]
-        self.assertEqual(value, nv)
-        self.assertTrue(isinstance(nv, type))
+        if isinstance(value, bytes) and type is str:
+            self.assertEqual(value.decode('utf-8'), nv)
+        else:
+            self.assertEqual(value, nv)
+        if type is not None:
+            self.assertTrue(isinstance(nv, type))
 
         return kd2
 
@@ -52,8 +56,14 @@ class TestKVS(unittest.TestCase):
     def test_set_string(self):
         self.set_and_check_context("string", "stuff", str)
 
+    def test_set_none(self):
+        self.set_and_check_context("none", None, None)
+
     def test_set_unicode(self):
         self.set_and_check_context("unicode", "\u32db \u263a \u32e1", str)
+
+    def test_set_bytes(self):
+        self.set_and_check_context("bytes", bytes.fromhex("deadbeef"), bytes)
 
     def test_set_list(self):
         self.set_and_check_context("list", [1, 2, 3, 4], list)
@@ -62,6 +72,9 @@ class TestKVS(unittest.TestCase):
         self.set_and_check_context(
             "dict", {"thing": "stuff", "other thing": "more stuff"}, dict
         )
+
+    def test_set_not_legal_json(self):
+        self.set_and_check_context("badjson", b"{", str)
 
     def test_exists_dir(self):
         with flux.kvs.get_dir(self.f) as kd:

--- a/t/python/t0005-kvs.py
+++ b/t/python/t0005-kvs.py
@@ -31,33 +31,33 @@ class TestKVS(unittest.TestCase):
         with flux.kvs.get_dir(self.f) as d:
             self.assertIsNotNone(d)
 
-    def set_and_check_context(self, key, value):
+    def set_and_check_context(self, key, value, type):
         kd = flux.kvs.KVSDir(self.f)
         kd[key] = value
         kd.commit()
         nv = kd[key]
         self.assertEqual(value, nv)
-        self.assertFalse(isinstance(nv, bytes))
+        self.assertTrue(isinstance(nv, type))
         return kd
 
     def test_set_int(self):
-        self.set_and_check_context("int", 10)
+        self.set_and_check_context("int", 10, int)
 
     def test_set_float(self):
-        self.set_and_check_context("float", 10.5)
+        self.set_and_check_context("float", 10.5, float)
 
     def test_set_string(self):
-        self.set_and_check_context("string", "stuff")
+        self.set_and_check_context("string", "stuff", str)
 
     def test_set_unicode(self):
-        self.set_and_check_context("unicode", "\u32db \u263a \u32e1")
+        self.set_and_check_context("unicode", "\u32db \u263a \u32e1", str)
 
     def test_set_list(self):
-        self.set_and_check_context("list", [1, 2, 3, 4])
+        self.set_and_check_context("list", [1, 2, 3, 4], list)
 
     def test_set_dict(self):
         self.set_and_check_context(
-            "dict", {"thing": "stuff", "other thing": "more stuff"}
+            "dict", {"thing": "stuff", "other thing": "more stuff"}, dict
         )
 
     def test_exists_dir(self):
@@ -79,7 +79,7 @@ class TestKVS(unittest.TestCase):
         self.assertTrue(flux.kvs.exists(self.f, "flagcheck"))
 
     def test_remove(self):
-        kd = self.set_and_check_context("todel", "things to delete")
+        kd = self.set_and_check_context("todel", "things to delete", str)
         del kd["todel"]
         kd.commit()
         with self.assertRaises(KeyError):
@@ -96,7 +96,7 @@ class TestKVS(unittest.TestCase):
             self.assertEqual(kd["dir"]["other_thing"], "dirstuff")
 
     def test_set_deep(self):
-        self.set_and_check_context("a.b.c.e.f.j.k", 5)
+        self.set_and_check_context("a.b.c.e.f.j.k", 5, int)
 
     def test_bad_init(self):
         with self.assertRaises(ValueError):

--- a/t/python/t0005-kvs.py
+++ b/t/python/t0005-kvs.py
@@ -31,7 +31,7 @@ class TestKVS(unittest.TestCase):
         with flux.kvs.get_dir(self.f) as d:
             self.assertIsNotNone(d)
 
-    def set_and_check_context(self, key, value, msg=""):
+    def set_and_check_context(self, key, value):
         kd = flux.kvs.KVSDir(self.f)
         kd[key] = value
         kd.commit()

